File: src/domain/recovery.py
"""Closed, provider-neutral Phase 6 recovery contracts. No prompts or error text."""

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class Frozen(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class FailureClass(StrEnum):
    MODEL_IDENTITY_MISMATCH = "MODEL_IDENTITY_MISMATCH"
    READ_TIMEOUT = "READ_TIMEOUT"
    PROVIDER_DEADLINE_EXCEEDED = "PROVIDER_DEADLINE_EXCEEDED"
    TASK_DEADLINE_EXCEEDED = "TASK_DEADLINE_EXCEEDED"
    PROVIDER_UNAVAILABLE = "PROVIDER_UNAVAILABLE"
    REMOTE_PROTOCOL_ERROR = "REMOTE_PROTOCOL_ERROR"
    RATE_LIMIT = "RATE_LIMIT"
    PROVIDER_5XX = "PROVIDER_5XX"
    OUTPUT_CONTRACT_FAILURE = "OUTPUT_CONTRACT_FAILURE"
    NON_RECOVERABLE_RUNTIME_FAILURE = "NON_RECOVERABLE_RUNTIME_FAILURE"
    RESEARCH_SEMANTIC_FAILURE = "RESEARCH_SEMANTIC_FAILURE"
    UNKNOWN = "UNKNOWN"


class Action(StrEnum):
    RETRY_SAME_ROUTE = "RETRY_SAME_ROUTE"
    SWITCH_MODEL = "SWITCH_MODEL"
    SWITCH_PROVIDER = "SWITCH_PROVIDER"
    CAPABILITY_CHECK = "CAPABILITY_CHECK"
    WAIT_AND_RETRY = "WAIT_AND_RETRY"
    REPLAN_TASK = "REPLAN_TASK"
    CORRECT_RESEARCH_PATH = "CORRECT_RESEARCH_PATH"
    FAIL_TASK = "FAIL_TASK"
    FAIL_RUN = "FAIL_RUN"


class Health(StrEnum):
    HEALTHY = "HEALTHY"
    DEGRADED = "DEGRADED"
    UNAVAILABLE = "UNAVAILABLE"
    UNKNOWN = "UNKNOWN"


class Capability(StrEnum):
    VERIFIED = "VERIFIED"
    CANDIDATE = "CANDIDATE"
    UNKNOWN = "UNKNOWN"
    UNSUPPORTED = "UNSUPPORTED"
    QUARANTINED = "QUARANTINED"


RouteId = Literal["teamorouter-sol", "teamorouter-luna", "teamorouter-terra", "mimo-direct"]
ModelId = Literal["gpt-5.6-sol", "gpt-5.6-luna", "gpt-5.6-terra", "mimo-v2.5", "mimo-v2.5-pro"]


class Scope(Frozen):
    run_id: str
    task_id: str
    object_id: str
    scheme_id: str
    task_profile: str
    agent_id: str
    operation_id: str = Field(
        default="specialist", min_length=1, max_length=256, pattern=r"^[A-Za-z0-9:_-]+$"
    )
    context_hash: str = Field(pattern=r"^sha256:[0-9a-f]{64}$")
    contract_hash: str = Field(pattern=r"^sha256:[0-9a-f]{64}$")

    @model_validator(mode="after")
    def exact_identity(self):
        if not self.task_id.startswith(self.run_id + ":"):
            raise ValueError("Recovery task must belong to exact Run")
        for field in (
            self.run_id,
            self.task_id,
            self.object_id,
            self.scheme_id,
            self.task_profile,
            self.agent_id,
        ):
            if not field or len(field) > 256 or any(c.isspace() for c in field):
                raise ValueError("Invalid recovery identity")
        return self


class FailureAssessment(Frozen):
    failure_class: FailureClass
    failure_stage: Literal["PROVIDER", "OUTPUT", "IDENTITY", "RESEARCH", "RUNTIME"]
    recoverable: bool


class RecoveryBudget(Frozen):
    max_total_attempts_per_task: int = Field(default=3, ge=1, le=4)
    max_same_route_attempts: int = Field(default=1, ge=1, le=2)
    max_model_fallbacks: int = Field(default=1, ge=0, le=2)
    max_cross_provider_switches: int = Field(default=1, ge=0, le=1)
    max_capability_checks: int = Field(default=1, ge=0, le=3)
    max_recovery_decisions: int = Field(default=5, ge=0, le=5)
    max_runtime_replans: int = Field(default=0, ge=0, le=0)
    max_total_recovery_time: float = Field(default=300, gt=0, le=300)
    max_total_recovery_cost: float | None = Field(default=None, ge=0)


class Candidate(Frozen):
    route: RouteId
    provider: Literal["teamorouter", "mimo"]
    model: ModelId
    authority_exists: bool
    health: Health = Health.UNKNOWN
    capability: Capability = Capability.UNKNOWN
    exhausted: bool = False
    eligible: bool = False
    next_allowed_action: Action = Action.FAIL_TASK
    capability_evidence_refs: tuple[str, ...] = Field(default=(), max_length=12)

    @model_validator(mode="after")
    def governed_route(self):
        expected = {
            "teamorouter-sol": ("teamorouter", {"gpt-5.6-sol"}),
            "teamorouter-luna": ("teamorouter", {"gpt-5.6-luna"}),
            "teamorouter-terra": ("teamorouter", {"gpt-5.6-terra"}),
            "mimo-direct": ("mimo", {"mimo-v2.5", "mimo-v2.5-pro"}),
        }
        provider, models = expected[self.route]
        if self.provider != provider or self.model not in models:
            raise ValueError("Route/provider/model identity mismatch")
        return self


class RecoveryDecision(Frozen):
    scope: Scope
    action: Action
    target_route: RouteId | None = None
    reason_code: Literal[
        "NEXT_GOVERNED_ROUTE",
        "CAPABILITY_REQUIRED",
        "NO_ALLOWED_ROUTE",
        "BUDGET_EXHAUSTED",
        "NONRECOVERABLE",
    ]
    confidence: float = Field(default=1, ge=0, le=1)
    evidence_refs: tuple[str, ...] = Field(default=(), max_length=12)


class RecoveryContext(Frozen):
    scope: Scope
    failure: FailureAssessment
    current_route: RouteId
    candidates: tuple[Candidate, ...] = Field(max_length=4)
    attempt_history: tuple[RouteId, ...] = Field(max_length=4)
    remaining_attempts: int = Field(ge=0, le=4)
    remaining_checks: int = Field(ge=0, le=3)
    remaining_decisions: int = Field(ge=0, le=5)
    remaining_seconds: float = Field(ge=0, le=300)
    dependency_state: Literal["CURRENT_TASK_RUNNING"] = "CURRENT_TASK_RUNNING"
    downstream_state: Literal["WAITING_FOR_CURRENT_TASK"] = "WAITING_FOR_CURRENT_TASK"
    scheme_requirement: Literal["UNCHANGED_EXACT_INTENT"] = "UNCHANGED_EXACT_INTENT"
    evidence_refs: tuple[str, ...] = Field(default=(), max_length=12)


class RecoveryEvidence(Frozen):
    record_id: str = Field(pattern=r"^REC-[0-9a-f-]{36}$")
    scope: Scope
    kind: Literal["ATTEMPT_STARTED", "ATTEMPT_COMPLETED", "DECISION", "TERMINAL"]
    attempt_id: str | None = Field(default=None, pattern=r"^ATT-[0-9a-f-]{36}$")
    attempt_number: int = Field(default=0, ge=0, le=4)
    route: RouteId | None = None
    provider: Literal["teamorouter", "mimo"] | None = None
    model: ModelId | None = None
    capability_check: bool = False
    outcome: Literal["STARTED", "PASS", "FAIL", "ALLOW", "DENY", "CANCELLED"]
    failure_class: FailureClass | None = None
    actual_model: ModelId | None = None
    requested_model: ModelId | None = None
    execution_policy: dict | None = None
    execution_outcome: (
        Literal[
            "MODEL_EXECUTION_DIRECT",
            "MODEL_EXECUTION_SUBSTITUTED",
            "MODEL_IDENTITY_OUT_OF_POLICY",
            "PROVIDER_IDENTITY_OUT_OF_POLICY",
        ]
        | None
    ) = None
    policy_gate_result: Literal["ALLOW", "DENY"] | None = None
    candidate_hash: str | None = Field(default=None, pattern=r"^sha256:[0-9a-f]{64}$")
    reason_code: Literal[
        "OBSERVED",
        "POLICY_DENIED",
        "RECOVERY_BUDGET_EXHAUSTED",
        "INTERRUPTED_EXECUTION",
        "NONRECOVERABLE",
    ] = "OBSERVED"
    action: Action | None = None
    decision: RecoveryDecision | None = None
    recovery_context: RecoveryContext | None = None
    timestamp: str
    latency_ms: float | None = Field(default=None, ge=0)
    input_tokens: int | None = Field(default=None, ge=0)
    output_tokens: int | None = Field(default=None, ge=0)
    output_hash: str | None = Field(default=None, pattern=r"^sha256:[0-9a-f]{64}$")

    @model_validator(mode="after")
    def validate_execution_policy(self):
        if self.execution_policy is not None:
            from src.domain.model_execution import ModelExecutionPolicy

            policy = ModelExecutionPolicy.model_validate(self.execution_policy)
            if (
                policy.task_profile != self.scope.task_profile
                or policy.provider_route != self.route
                or policy.preferred_model != self.requested_model
                or policy.provider != self.provider
            ):
                raise ValueError("Execution policy scope mismatch")
            if self.outcome == "PASS":
                outcome = policy.outcome(self.provider, self.requested_model, self.actual_model)
                if (
                    outcome not in {"MODEL_EXECUTION_DIRECT", "MODEL_EXECUTION_SUBSTITUTED"}
                    or outcome != self.execution_outcome
                    or self.policy_gate_result != "ALLOW"
                ):
                    raise ValueError("Execution identity is outside policy")
        return self

File: apps/web/src/types/recovery.ts
/** Safe display projection of the additive backend recovery ledger. */
export interface RecoveryRow {
  id: string; runId: string; taskId: string; profile: string; kind: string;
  attempt: number; check: boolean; route: string | null; model: string | null;
  outcome: string; action: string | null; failure: string | null; reason: string;
  latencyMs: number | null;
  preferredModel?: string | null; actualModel?: string | null; routing?: string | null;
  policyGate?: string | null;
}
const kinds = ["ATTEMPT_STARTED", "ATTEMPT_COMPLETED", "DECISION", "TERMINAL"];
const actions = ["RETRY_SAME_ROUTE", "SWITCH_MODEL", "SWITCH_PROVIDER", "CAPABILITY_CHECK", "WAIT_AND_RETRY", "REPLAN_TASK", "CORRECT_RESEARCH_PATH", "FAIL_TASK", "FAIL_RUN"];
const failures = ["MODEL_IDENTITY_MISMATCH", "READ_TIMEOUT", "PROVIDER_UNAVAILABLE", "REMOTE_PROTOCOL_ERROR", "RATE_LIMIT", "PROVIDER_5XX", "OUTPUT_CONTRACT_FAILURE", "NON_RECOVERABLE_RUNTIME_FAILURE", "RESEARCH_SEMANTIC_FAILURE", "UNKNOWN"];
function record(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error("Invalid recovery record");
  return value as Record<string, unknown>;
}
function choice(value: unknown, allowed: readonly string[]): string {
  if (typeof value !== "string" || !allowed.includes(value)) throw new Error("Invalid recovery code");
  return value;
}
function identity(value: unknown): string {
  if (typeof value !== "string" || !/^[A-Za-z0-9_:.-]{1,256}$/.test(value)) throw new Error("Invalid recovery identity");
  return value;
}
export function decodeRecoveryEvidence(value: unknown, runId: string, objectId: string): readonly RecoveryRow[] {
  if (!Array.isArray(value) || value.length > 10000) throw new Error("Invalid recovery collection");
  const ids = new Set<string>();
  let scheme: string | undefined;
  return value.map(raw => {
    const r = record(raw), s = record(r.scope);
    if (s.run_id !== runId || s.object_id !== objectId) throw new Error("Recovery identity mismatch");
    const taskId = identity(s.task_id), schemeId = identity(s.scheme_id), id = identity(r.record_id);
    if (!taskId.startsWith(runId + ":") || !/^REC-[0-9a-f-]{36}$/.test(id) || ids.has(id)) throw new Error("Recovery task/record mismatch");
    ids.add(id);
    scheme ??= schemeId;
    if (scheme !== schemeId) throw new Error("Recovery Scheme mismatch");
    const kind = choice(r.kind, kinds);
    const models = ["gpt-5.6-sol", "gpt-5.6-luna", "gpt-5.6-terra", "mimo-v2.5", "mimo-v2.5-pro"];
    const route = r.route === null ? null : choice(r.route, ["teamorouter-sol", "teamorouter-luna", "teamorouter-terra", "mimo-direct"]);
    const model = r.model === null ? null : choice(r.model, models);
    if (kind.startsWith("ATTEMPT_")) {
      const pair = route === "teamorouter-sol" ? ["teamorouter", "gpt-5.6-sol"] : route === "teamorouter-luna" ? ["teamorouter", "gpt-5.6-luna"] : route === "teamorouter-terra" ? ["teamorouter", "gpt-5.6-terra"] : ["mimo", model];
      if (!route || !model || r.provider !== pair[0] || model !== pair[1] || (route === "mimo-direct" && !model.startsWith("mimo-"))) throw new Error("Recovery route mismatch");
    }
    const action = r.action === null ? null : choice(r.action, actions);
    if (kind === "DECISION") {
      const decision = record(r.decision), scope = record(decision.scope);
      if (scope.run_id !== runId || scope.object_id !== objectId || scope.task_id !== taskId || scope.scheme_id !== schemeId || decision.action !== action || decision.target_route !== route) throw new Error("Recovery decision mismatch");
    }
    if (typeof r.capability_check !== "boolean" || !Number.isInteger(r.attempt_number) || Number(r.attempt_number) < 0 || Number(r.attempt_number) > 4) throw new Error("Invalid recovery attempt");
    if (r.latency_ms !== null && (typeof r.latency_ms !== "number" || !Number.isFinite(r.latency_ms) || r.latency_ms < 0)) throw new Error("Invalid recovery latency");
    const policy = r.execution_policy == null ? null : record(r.execution_policy);
    const actualModel = r.actual_model == null ? null : choice(r.actual_model, models);
    const routing = r.execution_outcome == null ? null : choice(r.execution_outcome, ["MODEL_EXECUTION_DIRECT", "MODEL_EXECUTION_SUBSTITUTED", "MODEL_IDENTITY_OUT_OF_POLICY", "PROVIDER_IDENTITY_OUT_OF_POLICY"]);
    if (policy) {
      if (policy.provider_route !== route || policy.provider !== r.provider || policy.preferred_model !== model || policy.task_profile !== s.task_profile || policy.version !== "bounded-model-execution/v1" || typeof policy.dynamic_substitution_allowed !== "boolean" || !Array.isArray(policy.authorized_actual_models)) throw new Error("Invalid model policy binding");
      const allowed = policy.authorized_actual_models.map(m => choice(m, models));
      if (r.outcome === "PASS" && (!actualModel || !allowed.includes(actualModel) || (actualModel !== model && (!policy.dynamic_substitution_allowed || routing !== "MODEL_EXECUTION_SUBSTITUTED")))) throw new Error("Execution outside policy");
    }
    return {
      id, runId, taskId, profile: choice(s.task_profile, ["fundamental_analysis", "peer_analysis", "research_news_analysis", "valuation_analysis", "risk_analysis", "risk_follow_up", "report_synthesis", "GENERATED_CAPABILITY:free_cash_flow_margin"]),
      kind, route, model, attempt: Number(r.attempt_number), check: r.capability_check,
      outcome: choice(r.outcome, ["STARTED", "PASS", "FAIL", "ALLOW", "DENY", "CANCELLED"]), action,
      failure: r.failure_class === null ? null : choice(r.failure_class, failures),
      reason: choice(r.reason_code, ["OBSERVED", "POLICY_DENIED", "RECOVERY_BUDGET_EXHAUSTED", "INTERRUPTED_EXECUTION", "NONRECOVERABLE"]),
      latencyMs: r.latency_ms as number | null,
      preferredModel: policy ? model : null, actualModel, routing,
      policyGate: r.policy_gate_result == null ? null : choice(r.policy_gate_result, ["ALLOW", "DENY"])
    };
  });
}
